### PR TITLE
Add to_flat_index method to MultiIndex

### LINKED ISSUE #19950: Add method to flatten all multi-index levels
Sometimes it is useful to flatten all levels of a multi-index. For example, when pivoting data into a wide format, the new columns are generally multi-indexed. However, when exporting to CSV, sometimes it might be desirable to have only one header row.

AFAIK, there is no dedicated method to flatten an existing multi-index. Assuming I want to combine the multi-index levels with an underscore, I would do this to get one header row:

```
df.columns = ["_".join(v) for v in df.columns.values]
````

This is not really obvious, and maybe a dedicated `MultiIndex` method should be added, e.g. `df.columns.flatten(sep="_", inplace=False)`. What do you think?

### PR DESCRIPTION
**Pull request by WillAyd**
- [X] closes #19950
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Very simple implementation at the moment. The thought here is to introduce this method and perhaps subsequently extend to allow for string concatenation of the elements. Longer term there could also be a keyword added to `.agg` of GroupBy which will dispatch to this instead of simply returning a MultiIndex column, which could alleviate some of the pain users are experience when trying to rename columns after an aggregation.

@TomAugspurger and @jorisvandenbossche from the dev chat today

### CODE REVIEW

**Comment by jreback** [comment]
can you add to api.rst. I would add the ``sep=None`` arg here, maybe can defer the actual handling of the string concat to a later PR

**Comment by WillAyd** [comment]
Yep my goal is to defer the actual sep implementation, if only because I'm not sure how we'd want to handle say timestamp and categorical objects so probably needs further discussion. 

For now I've stubbed it out and raise a `NotImplementedError` but if you want to go about in a different way lmk

**Comment by jorisvandenbossche** [comment]
> I'm not sure how we'd want to handle say timestamp and categorical objects so probably needs further discussion.

I think if you specify `sep`, everything gets stringified? (just using the default conversion to string, without any ability to use a custom format, the user can always do that before they use this method)

---

About the name, do we want to call this `to_index` ? It might feel a bit strange, since you already have an index to start with (it's only a MultiIndex, but which is still an index). 
So you would get something like `df.index.to_index()` or `df.columns.to_index()`, which seems a bit strange.

Alternative would be `.flatten(..)`, but I don't know to what extent it is a problem that this clashes with the numpy method. 
Or `to_single`, `collapse_levels`, ... (but I don't really like those :-))

**Comment by TomAugspurger** [comment]
Yesterday, I almost left the same comment as Joris about the name :)
Something about MultiIndex.to_index bothered
me, because it implies that MultiIndex isn't an Index.

In the end, I deleted the comment because

1.) there's a nice symmetry with the other to_* methods
2.) I couldn't come up with a better name.

**Comment by WillAyd** [comment]
Agreed with all comments so far - the name was initially off-putting to me but ultimately went with it given the consistency of it with the API, and I didn't want to obfuscate what `flatten` meant from a NumPy perspective.

> I think if you specify `sep`, everything gets stringified?

Probably. I'm probably being ultra-conservative here but something about the string conversion and "round-tripability" of this is why I was planning on holding off until a separate PR to even go down that path. If it's a blocker here can certainly add the simple implementation.

> `Index.to_index()` tested? (iow identity)

No but I also don't believe this is implemented atm. Are you thinking this implementation in this PR is better served in the `base.py` module?

**Comment by WillAyd** [comment]
Any other thoughts on this?

File: pandas/core/indexes/multi.py [hunk 1]
```diff
@@ -1198,6 +1199,28 @@ def to_frame(self, index=True, name=None):
             result.index = self
         return result
```
Comment (jreback) [inline pandas/core/indexes/multi.py]: how about to_flat ?

Comment (jreback) [reply pandas/core/indexes/multi.py]: also i would add the identity method to Index itself

**Comment by WillAyd** [comment]
Added the identity method. I did not change the name from `to_index` to `to_flat` because I personally prefer the API consistency with `to_series` and `to_frame` though I'm not going to get terribly hung up on that if we feel `to_flat` is better.

I also kept the documentation geared only towards MultiIndex atm. Not sure if we want to document the identity and if it serves a purpose outside of not breaking chained ops (at least in its current state) but happy to update with whatever we feel necessary

**Comment by jorisvandenbossche** [comment]
As said above, I am also -1 on `to_index`, this is just a confusing name. I agree with @shoyer that it's then better to have a verbose but at least readable / clear name. 

Other names I was thinking about: `.flatten_levels()` or `flatten_index_levels` (this last one is maybe too verbose, although explicit). But `to_flat_index()` is fine for me.

Now you added it to the Index/MultiIndex. But if you want to do be able to directly do this is in a method chain (eg after a groupby operation), it would be needed on the DataFrame as well. Is this something we would also want to consider (but perfectly fine with start to add it at least to the index itself)

**Comment by toobaz** [comment]
Sorry for coming late. I'm also not convinced by ``to_index``, and weakly in favor of ``to_flat`` (although ``flatten`` would not disturb me).

I'm also not convinced by ``sep``. I think that:
- the necessity that emerged lately (e.g. in the mailing list, when discussing ``MultiIndex``es produced by ``GroupBy.apply``) was for instance to transform a level of operation names ('mean', 'min', 'max') and one of variables ('a', 'b', 'c') in an index of strings such as 'mean of a', 'mean of b' and so on. Now, ``sep=' of '`` does (once implemented) allow this, but just because it happens that there are two levels, and in the correct order, not because it makes much sense to think to the labels for each level as a flat ordered list
- I think giving less freedom today in terms of aggregation methods just calls for adding more arguments later

My idea would be to pass a callable (e.g. ``fmt``) accepting an iterable and returning anything (not necessarily a string, by the way). The ``sep`` solution would just become, on a ``MultiIndex`` with 2 levels, ``fmt='{} of {}'.format``.

In any case, if we want to discuss this and not block this PR, I would just remove the ``sep`` argument for the moment.

**Comment by jorisvandenbossche** [comment]
> My idea would be to pass a callable (e.g. fmt) accepting an iterable and returning anything

The idea is that this would be a keyword of this (to be renamed) `to_index` method, or a method on its own?

In any case, I like this idea, it gives indeed more flexibility than a `sep`.

We had some discussion about this before (I think in light of deprecating `add_prefix`/`add_suffix`) to have an easier way to format values in an Index/Series (related issues: https://github.com/pandas-dev/pandas/pull/18347, https://github.com/pandas-dev/pandas/issues/17211)

**Comment by jorisvandenbossche** [comment]
This is kind of 

```
In [142]: idx = pd.MultiIndex.from_product([['mean', 'sum'], ['a', 'b']])

In [143]: idx.map(lambda x: "{0} of {1}".format(*x))
Out[143]: Index(['mean of a', 'mean of b', 'sum of a', 'sum of b'], dtype='object')
```

File: pandas/core/indexes/base.py [hunk 2]
```diff
@@ -1105,6 +1105,52 @@ def _format_attrs(self):
...
+              dtype='object')
+        """
+        if sep is not None:
+            # TODO: Add support for separator to return strs instad of tuples
+            raise NotImplementedError
+
+        if not isinstance(self, ABCMultiIndex):
+            return self
+
+        idx = Index(self.values, tupleize_cols=False)
```
Comment (toobaz) [inline pandas/core/indexes/base.py]: Maybe ``iter(self)`` is slightly more efficient? (as I think it happens anyway inside ``self.values``, and I don't think passing through an ``array`` helps)

**Comment by jreback** [comment]
after reading #23141 maybe we should call this ``.squeeze()``?

**Comment by jreback** [comment]
from https://github.com/pandas-dev/pandas/pull/22866#issuecomment-437688069

ahh right, ok then.

**Comment by toobaz** [comment]
> This is a bit different from the squeeze semantics.
> 
> squeeze will reduce dimensionality _if possible_ (only one item on that axis).
> 
> `.flatten` / `.to_index` / whatever will always reduce the dimensionality.

Just for pickiness: the two differ in fact also for the "only one item on that axis" case, as the former will return the labels as items, while the latter will return length one tuples as items.

**Comment by WillAyd** [comment]
Renamed to `to_flat_index` and removed `sep` parameter based off feedback above

**Comment by toobaz** [comment]
Seems good to me

**Review by TomAugspurger — APPROVED** [review]
Shall we merge?

File: pandas/core/indexes/base.py [hunk 3]
```diff
@@ -1113,6 +1113,39 @@ def _format_attrs(self):
...
+        --------
+        >>> index = pd.MultiIndex.from_product(
+        ...     [['foo', 'bar'], ['baz', 'qux']],
+        ...     names=['a', 'b'])
+        >>> index.to_flat_index()
+        Index([('foo', 'baz'), ('foo', 'qux'),
+               ('bar', 'baz'), ('bar', 'qux')],
+              dtype='object')
+        """
+        if not isinstance(self, ABCMultiIndex):
```
[the remark this answers is not in the document]
Comment (toobaz) [reply pandas/core/indexes/base.py]: Preference for 2, and in general, for leaving the base ``Index`` class unaware (every time it is possible/reasonable) of the existence and specificities of ``Index`` subclasses.

File: pandas/core/indexes/base.py (hunk 2 above)
Comment (toobaz) [reply pandas/core/indexes/base.py]: Sorry, I had forgot this... what do you @WillAyd think?

Comment (TomAugspurger) [reply pandas/core/indexes/base.py]: If use don't do it here, the `Index` constructors has to convert tot an array anyway for storage, right?

```python
In [4]: pd.Index(np.array([(1, 2), (3, 4)], dtype=object), tupleize_cols=False).values
Out[4]: array([(1, 2), (3, 4)], dtype=object)
```

In theory, the array of tuples should just pass through without copy.

**Comment by WillAyd** [comment]
The inheritance model does seem to make more sense here so I've updated the latest commit to reflect that

File: pandas/core/indexes/base.py (hunk 2 above)
Comment (toobaz) [reply pandas/core/indexes/base.py]: @TomAugspurger You're right

``` python
In [2]: a = np.array([1, 2], dtype=object)

In [3]: a[:] = [(1, 2), (3, 4)]

In [4]: id(a), id(pd.Index(a, tupleize_cols=False).values.base)
Out[4]: (140686849540864, 140686849540864)
```
sorry for the noise

**Comment by TomAugspurger** [comment]
Thanks @WillAyd!

File: pandas/core/indexes/base.py [hunk 4]
```diff
@@ -1113,6 +1113,26 @@ def _format_attrs(self):
         """
         return format_object_attrs(self)
 
+    def to_flat_index(self):
+        """
+        Identity method.
+
+        .. versionadded:: 0.24.0
+
+        This is implemented for compatability with subclass implementations
```
Comment (toobaz) [inline pandas/core/indexes/base.py]: compatability -> compatibility

By the way: reusing the same docstring would be a + for me. I would rather describe the actual operation (so on ``MuiltiIndex``) and then say that it is idempotent on flat ``Index``es.

**Pull request merged**